src/enzo/ZeusFDM.C

```diff
@@ -0,0 +1,210 @@
+/***********************************************************************
+c
+c  SOURCE TERMS FOR ZEUS HYDRO (CARTESIAN ONLY)
+c
+c  written by: Greg Bryan (implemented from Stone & Norman, ApJS 80, 753)
+c  date:       February, 1997
+c  modified1: 
+c
+c  PURPOSE:
+c     Adds the source substeps
+c
+c  EXTERNALS:
+c
+c  INPUTS:
+c     d       - density field (includes boundary zones)
+c     dx,y,z  - zone width arrays for each dimension
+c     e       - total specific energy field
+c     ge      - gas energy (used when idual = 1)
+c     gr_x,y,zacc - gravitational acceleration fields
+c     gravity - flag indicating whether or not to use gravity field (1 = yes)
+c     i,j,kn  - dimensions of field arrays
+c     igamfield - indicates if gamma should be a field
+c     ipresfree - pressure-free flag (0 = off, 1 = on, i.e. p=0)
+c     rank    - dimension of problem (not currently used)
+c     u       - x-velocity field
+c     v       - y-velocity field
+c     w       - z-velocity field
+c     C1,C2   - Linear and quadratic artifificla viscosity parameters
+c     minsupecoef - coefficient for minimum pressure support
+c
+************************************************************************/
+
+#include <stdio.h>
+#include <math.h>
+#include "ErrorExceptions.h"
+#include "macros_and_parameters.h"
+#include "fortran.def"
+
+#define IDX(a,b,c) ( ((c)*jn + (b))*in + (a) )
+
+int ZeusFDM(float *d, float *e, float *u, float *v, float *w, float *p,
+	       int in, int jn, int kn, int rank,
+	       int is, int ie, int js, int je, int ks, int ke, 
+	       float C1, float C2, float *gamma, float dt, float dx[], float dy[], float dz[],
+	       int gravity, float *gr_xacc, float *gr_yacc, float *gr_zacc, 
+	       float minsupecoef, float lapcoef)
+{
+  int ijk = MAX_ANY_SINGLE_DIRECTION;
+
+  /* Local declarations */
+
+  int i, j, k, jsm1, ksm1, jep1, kep1, ism2, jsm2, ksm2, jep2, kep2, n, nstep;
+  int size = in*jn*kn;
+  float gamma1 = gamma[0];
+  float  deltav, dt1, e1;
+  float *logd = new float[size];
+  float q[ijk];
+
+  /* ======================================================================= */
+
+  /* Compute varients on start indexes */
+
+  jsm1 = max(js-1, 0);
+  jsm2 = max(js-2, 0);
+  jep1 = min(je+1, jn-1);
+  jep2 = min(je+2, jn-1);
+  ksm1 = max(ks-1, 0);
+  ksm2 = max(ks-2, 0);
+  kep1 = min(ke+1, kn-1);
+  kep2 = min(ke+2, kn-1);
+
+  /* compute log of density */
+  
+   for (i = 0; i < size; i++) {
+    if (d[i] < 0) {
+      fprintf(stderr, "u,v,w,d,e=%"GSYM",%"GSYM",%"GSYM",%"GSYM",%"GSYM"  dx=%"GSYM"  dt=%"GSYM"\n", 
+        u[i],v[i],w[i],d[i],p[i], dx[0], dt);
+      ENZO_FAIL("ZeusSolver: Negative Density! \n");
+    } else{
+      logd[i] = log(d[i]);
+    }
+  }
+
+  /* Compute the quantum pressure */
+
+#define QP_4TH_ORDER
+
+  for (k = 0; k < kn; k++) {
+    for (j = 0; j < jn; j++) {
+      for (i = 0; i < in; i++){
+
```

Comment:
Not a necessary change, but there is a lot of repeated calculation through all of the `IDX` calls.

Reply:
Yes, maybe the compiler can optimized -- I added a comment, but didn't fixed at this point.